### PR TITLE
Fix 77603 - Don't break the log line when it is more than 16K

### LINKED ISSUE #77603: Pod logs: long lines are corrupted when using timestamps=true
**What happened**:

Docker has a hardcoded 16K limit on log line length, which splits lines longer than that into multiple messages in the log file. `ReadLogs` in [logs.go](https://github.com/kubernetes/kubernetes/blob/255a73ffc64d1c4e0f9a378ebf8d0005edfe6962/pkg/kubelet/kuberuntime/logs/logs.go), however, doesn't account for this when `timestamps` is set to `true` and prints the timestamp in front of every entry in the log file instead of every logical line. This leads to random timestamps appearing in the middle of log lines, potentially breaking the applications that parse these logs.

**What you expected to happen**:

`ReadLogs` inserts timestamps once per logical line, so original log data is preserved. Alternatively, allow `/logs` API to request the logs in a sane format like JSON, which would allow merging the lines back and combining them with the timestamps more intelligently on the application side.

**How to reproduce it (as minimally and precisely as possible)**:

* Create `foo.yaml`:

      apiVersion: v1
      kind: Pod
      metadata:
        name: log-lines
      spec:
        containers:
        - name: main
          image: busybox
          command: ['sh', '-c', 'while sleep 1; do echo -n ___; for i in $(seq 1 18000); do echo -n A; done; echo ___; done']

 * Start the pod with `kubectl apply -f foo.yaml`
 * Fetch the logs with `kubectl logs log-lines --timestamps`, or using the API (`curl https://…/api/v1/namespaces/default/pods/log-lines/log?timestamps=true&container=main`)
 * Observe random timestamps inserted in the middle of the log lines.

**Environment**:
- Kubernetes version (use `kubectl version`):
```Client Version: version.Info{Major:"1", Minor:"14", GitVersion:"v1.14.1", GitCommit:"b7394102d6ef778017f2ca4046abbaa23b88c290", GitTreeState:"clean", BuildDate:"2019-04-19T22:12:47Z", GoVersion:"go1.12.4", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.5", GitCommit:"2166946f41b36dea2c4626f90a77706f426cdea2", GitTreeState:"clean", BuildDate:"2019-03-25T15:19:22Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

### PR DESCRIPTION
**Pull request by EricWvi**
#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
When the log line's content size is more than 16K, "kubectl logs --timestamps" breaks the line as multiple lines .

It is not convenient for other tool to parse the log
#### Which issue(s) this PR fixes:

Fixes #77603

#### release note
```release-note
When container log's length exceeds 16K, 
Docker will split the log entry into two entries. 
One contains "\n", the other not, 
which may affect log processing tools.

This fix also takes --tail, --timestamps, -f into account.
```

/cc @SergeyKanzhelev

### CODE REVIEW

File: pkg/kubelet/kuberuntime/logs/logs_test.go [hunk 1]
```diff
@@ -72,14 +74,23 @@ func TestLogOptions(t *testing.T) {
...
+		`{"log":"line2\n","stream":"stdout","time":"2020-09-27T11:18:02.00000000Z"}` + "\n",
+		`{"log":"line3\n","stream":"stdout","time":"2020-09-27T11:18:03.00000000Z"}` + "\n",
+		`{"log":"line4 part I ","stream":"stdout","time":"2020-09-27T11:18:04.00000000Z"}` + "\n",
+		`{"log":"+ II ","stream":"stdout","time":"2020-09-27T11:18:04.00000000Z"}` + "\n",
+		`{"log":"+ III\n","stream":"stdout","time":"2020-09-27T11:18:04.00000000Z"}` + "\n",
+		`{"log":"line5\n","stream":"stdout","time":"2020-09-27T11:18:05.00000000Z"}` + "\n",
+		`{"log":"line6\n","stream":"stdout","time":"2020-09-27T11:18:06.00000000Z"}` + "\n",
+	}
+
+	realTimeLog := []string{
```
Comment (EricWvi) [inline pkg/kubelet/kuberuntime/logs/logs_test.go:88]: Add test for `kubectl logs -f`

**Comment by pacoxu** [comment]
/priority important-longterm
/cc neolit123
as you have reviewed #77822

**Comment by neolit123** [comment]
i'm going to leave this one to sig-instrumentation / node maintainers to review.
/sig instrumentation

**Comment by dashpole** [comment]
We did a quick review of this during triage.  Does this only impact kubectl logs when specifying `--timestamp`.  Also, the linked issue is already closed.  That may need to be reopened and discussed if it wasn't already fixed.  Does this only impact kubectl logs, or does this change how logs are written to disk?

/assign @ehashman

**Comment by ehashman** [comment]
/hold

I am not sure this is the right way to solve this issue, per @dashpole's comments above (we discussed during SIG Instrumentation triage)

**Comment by EricWvi** [comment]
@dashpole 
Yeah, I wanted to re-open the issue, but failed. Thanks for @pacoxu , now it is open.
The problem has not been fixed so far and I reproduced it in the cluster with Kubernetes 1.21.
```
Client Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T21:04:39Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"21", GitVersion:"v1.21.3", GitCommit:"ca643a4d1f7bfe34773c74f79527be4afd95bf39", GitTreeState:"clean", BuildDate:"2021-07-15T20:59:07Z", GoVersion:"go1.16.6", Compiler:"gc", Platform:"linux/amd64"
```
#### Pod Details
- one api server developed by gin

#### Scene
- I deliberately insert one record into the database, which is a little bit longer than 16K.
- I make a request for a document referring to that record.
- `k logs --tail=8 --prefix --timestamps dockerlog` (I highlight the prefix and timestamps)
image
- ”should halt an entire program or server request“ is the end of the document. It is splitted from the last log.
- `k logs --tail=9 --prefix --timestamps dockerlog`, the timestamp is inserted between two logs. Besides, as for one piece of log, it is splitted into two. 
image
...
image

**Comment by EricWvi** [comment]
I think that it indeed only impact kubectl logs, rather than logs written to disk. In this scene, the logs written to disk are produced by Docker. Logs in `/var/log/containers` are symlinks.
I'm all ears.

**Pull request closed without merging**